### PR TITLE
Fixed Java Core issue 1585 - ForestDB + Encryption - SIGSEGV during d…

### PR DESCRIPTION
**Pull request by hideki**
…atabase deletion

Issue: https://github.com/couchbase/couchbase-lite-java-core/issues/1585

Two issues
- JNI does not allow to pass array longer than 512 items. Separate the query if doc IDs are more than 500.
- FDB_RESULT_HANDLE_BUSY is thrown from cbforest if forestdb is busy. Implement retry logic in case DB Busy is thrown from forestdb.

### CODE REVIEW

**Comment by hideki** [comment]
@pasin If you have time, please review this PR?

File: src/main/java/com/couchbase/lite/store/ForestDBStore.java [hunk 1]
```diff
@@ -644,105 +648,147 @@ public int compare(String id1, String id2) {
...
-                            && doc.conflicted()) {
-                        conflicts = ForestBridge.getCurrentRevisionIDs(doc);
-                        if (conflicts != null && conflicts.size() == 1)
-                            conflicts = null;
+                    DocumentIterator itr = forest.iterator(docIDs, iteratorFlags);
+                    try {
+                        List<QueryRow> retRows = readFromIterator(itr, options, includeDocs, filter, limit);
+                        rows.addAll(retRows);
+                        limit -= retRows.size();
+                        total -= retRows.size();
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/store/ForestDBStore.java:665]: The while loop `while (total > 0)` in line 656 will keep looping if the total value is more than zero. If case of the `retRows` is less than `MAX_RECORDS_TO_READ_FROM_FORESTDB_AT_ONCE` (e.g. some doc ids don't exists in the database), could it cause the infinite loop? 

Should it continue looping if there are still some docIDs left for querying and the total rows are still less than limit?

Comment (hideki) [reply src/main/java/com/couchbase/lite/store/ForestDBStore.java:665]: > (e.g. some doc ids don't exists in the database)

In this case, the row should have `missing doc` status.

> could it cause the infinite loop?

There is a possibility that storage could return incorrect results. In this case, it causes the infinite loop. Maybe it is better to exit the loop if remaining docIDs are empty

> Should it continue looping if there are still some docIDs left for querying and the total rows are still less than limit?

not necessary

Comment (hideki) [reply src/main/java/com/couchbase/lite/store/ForestDBStore.java:665]: Filed new ticket for this: https://github.com/couchbase/couchbase-lite-java-core/issues/1591

Comment (pasin) [reply src/main/java/com/couchbase/lite/store/ForestDBStore.java:665]: > In this case, the row should have missing doc.

What is about the case that the document was deleted? Will it is listed as missing or deleted row?

Comment (hideki) [reply src/main/java/com/couchbase/lite/store/ForestDBStore.java:665]: https://github.com/couchbaselabs/couchbase-lite-java-forestdb/pull/93/files#diff-f38e334f05bac26d7c898fc48f6f78d5R729
```
!exist -> missing doc
!Flag-includedeleted & deleted -> no row
Flag-conflict-only & !conflicted -> no row
```
some corner cases....

Comment (hideki) [reply src/main/java/com/couchbase/lite/store/ForestDBStore.java:665]: PR: https://github.com/couchbaselabs/couchbase-lite-java-forestdb/pull/94

**Pull request merged**